src/framework/storage/FileSystemGraphObjectStore/FileSystemGraphObjectStore.ts

```diff
@@ -0,0 +1,138 @@
+import pMap from 'p-map';
+import { Sema } from 'async-sema';
+
+import { Entity, Relationship } from '../../types';
+import { GraphObjectFilter, GraphObjectIteratee } from '../../execution/types';
+
+import { flushDataToDisk } from './flushDataToDisk';
+import { BucketMap } from './BucketMap';
+
+import {
+  iterateEntityTypeIndex,
+  iterateRelationshipTypeIndex,
+} from './indices';
+
+export const GRAPH_OBJECT_BUFFER_THRESHOLD = 500; // arbitrarily selected, subject to tuning
+
+interface FileSystemGraphObjectStoreInput {
+  cacheDirectory?: string;
+}
+
+export class FileSystemGraphObjectStore {
+  readonly cacheDirectory?: string;
+
+  semaphore: Sema;
+  entityStorageMap: BucketMap<Entity>;
+  relationshipStorageMap: BucketMap<Relationship>;
+
+  constructor(options?: FileSystemGraphObjectStoreInput) {
+    this.cacheDirectory = options?.cacheDirectory;
+
+    this.entityStorageMap = new BucketMap();
+    this.relationshipStorageMap = new BucketMap();
+
+    this.semaphore = new Sema(1);
+  }
+
+  async addEntities(storageDirectoryPath: string, newEntities: Entity[]) {
+    this.entityStorageMap.add(storageDirectoryPath, newEntities);
+
+    if (this.entityStorageMap.totalItemCount >= GRAPH_OBJECT_BUFFER_THRESHOLD) {
+      await this.flushEntitiesToDisk();
+    }
+  }
+
+  async addRelationships(
+    storageDirectoryPath: string,
+    newRelationships: Relationship[],
+  ) {
+    this.relationshipStorageMap.add(storageDirectoryPath, newRelationships);
+
+    if (
+      this.relationshipStorageMap.totalItemCount >=
+      GRAPH_OBJECT_BUFFER_THRESHOLD
+    ) {
+      await this.flushRelationshipsToDisk();
+    }
+  }
+
+  async iterateEntities(
+    filter: GraphObjectFilter,
+    iteratee: GraphObjectIteratee<Entity>,
+  ) {
+    await this.flushEntitiesToDisk();
+
+    await iterateEntityTypeIndex({
+      cacheDirectory: this.cacheDirectory,
+      type: filter._type,
+      iteratee,
+    });
+  }
+
+  async iterateRelationships(
+    filter: GraphObjectFilter,
+    iteratee: GraphObjectIteratee<Relationship>,
+  ) {
+    await this.flushRelationshipsToDisk();
+
+    await iterateRelationshipTypeIndex({
+      cacheDirectory: this.cacheDirectory,
+      type: filter._type,
+      iteratee,
+    });
+  }
+
+  async flush() {
+    await Promise.all([
+      this.flushEntitiesToDisk(),
+      this.flushRelationshipsToDisk(),
+    ]);
+  }
+
+  async flushEntitiesToDisk() {
+    await this.lockOperation(
```

Comment:
I think it'd be great to get some of the commentary you added to the commit for this semaphore into this class's documentation. That was super helpful for explaining the purpose.

Reply:
Will add.